rest/route/project.go

```diff
@@ -883,83 +884,205 @@ func (h *getProjectVersionsHandler) Parse(ctx context.Context, r *http.Request)
 	}
-	if h.opts.Limit < 1 {
-		return errors.New("limit must be a positive integer")
+	if opts.Limit < 1 {
+		return nil, errors.New("limit must be a positive integer")
 	}
 
 	startStr := params.Get("start")
 	if startStr != "" {
 		startOrder, err := strconv.Atoi(params.Get("start"))
 		if err != nil {
-			return errors.Wrap(err, "invalid start query parameter")
+			return nil, errors.Wrap(err, "invalid start query parameter")
 		}
-		h.opts.StartAfter = startOrder
+		opts.StartAfter = startOrder
 	}
-	if h.opts.StartAfter < 0 {
-		return errors.New("start must be a non-negative integer")
+	if opts.StartAfter < 0 {
+		return nil, errors.New("start must be a non-negative integer")
 	}
 
-	requester := params.Get("requester")
-	if requester != "" {
-		h.opts.Requester = requester
+	endStr := params.Get("end")
+	if endStr != "" {
+		endOrder, err := strconv.Atoi(params.Get("end"))
+		if err != nil {
+			return nil, errors.Wrap(err, "invalid end query parameter")
+		}
+		opts.EndAt = endOrder
 	}
-	if h.opts.Requester == "" {
-		h.opts.Requester = evergreen.RepotrackerVersionRequester
+	if opts.EndAt < 0 {
+		return nil, errors.New("end must be a non-negative integer")
+	}
+	if opts.EndAt > opts.StartAfter {
+		return nil, errors.New("end must be less than or equal to start")
 	}
-	return nil
-}
 
-func (h *getProjectVersionsHandler) Run(ctx context.Context) gimlet.Responder {
-	versions, err := data.GetProjectVersionsWithOptions(h.projectName, h.opts)
-	if err != nil {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err, "getting versions for project '%s'", h.projectName))
+	if opts.StartTimeStr == "" && opts.StartAfter == 0 {
+		return nil, errors.New("must specify either timestamps or order numbers")
+	}
+	if opts.StartTimeStr != "" && opts.StartAfter != 0 {
+		return nil, errors.New("cannot specify both timestamps and order numbers")
 	}
 
-	resp, err := gimlet.NewBasicResponder(http.StatusOK, gimlet.JSON, versions)
-	if err != nil {
-		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "constructing response"))
+	var err error
+	if opts.StartTimeStr != "" {
+		opts.StartTime, opts.EndTime, err = getStartAndEndTime(opts.StartTimeStr, opts.EndTimeStr)
+		if err != nil {
+			return nil, errors.Wrap(err, "getting start and end time")
+		}
 	}
 
-	if len(versions) >= h.opts.Limit {
-		err = resp.SetPages(&gimlet.ResponsePages{
-			Next: &gimlet.Page{
-				Relation:        "next",
-				LimitQueryParam: "limit",
-				KeyQueryParam:   "start",
-				BaseURL:         h.url,
-				Key:             strconv.Itoa(versions[len(versions)-1].Order),
-				Limit:           h.opts.Limit,
-			},
-		})
+	requester := params.Get("requester")
+	if requester != "" {
+		opts.Requester = requester
+	}
+	if opts.Requester == "" {
+		opts.Requester = evergreen.RepotrackerVersionRequester
+	}
+	return opts, nil
+}
 
+func getStartAndEndTime(startTimeStr, endTimeStr string) (start time.Time, end time.Time, err error) {
```

Comment:
Can we remove the names from these return values?

Reply:
Sure. I only added them since we're returning two `Time` structs and thought it'd be good to be clearer which is which. But if it's a pattern we'd like to avoid for consistency that's fine.

Reply:
That logic makes sense -- in that case maybe a function comment?